packages/integrations/cloudflare/src/index.ts

```diff
@@ -105,17 +206,49 @@ export default function createIntegration(args?: Options): AstroIntegration {
 				_buildConfig = config.build;
 
 				if (config.output === 'static') {
-					throw new Error(`
-  [@astrojs/cloudflare] \`output: "server"\` or \`output: "hybrid"\` is required to use this adapter. Otherwise, this adapter is not necessary to deploy a static site to Cloudflare.
-
-`);
+					throw new Error(STATIC_ERROR);
 				}
 
 				if (config.base === SERVER_BUILD_FOLDER) {
-					throw new Error(`
-  [@astrojs/cloudflare] \`base: "${SERVER_BUILD_FOLDER}"\` is not allowed. Please change your \`base\` config to something else.`);
+					throw new Error(FOLDER_ERROR);
 				}
 			},
+			'astro:server:setup': ({ server }) => {
+				server.middlewares.use(async function middleware(req, res, next) {
+					if (runtimeMode === 'off') {
+						next();
+					} else {
+						const vars = await CFVars();
+						const cf = await getCFObject(runtimeMode);
+
+						const clientLocalsSymbol = Symbol.for('astro.locals');
+						Reflect.set(req, clientLocalsSymbol, {
+							runtime: {
+								env: {
+									// default binding for static assets will be dynamic once we support mocking of bindings
+									ASSETS: {},
+									// this is just a VAR for CF to change build behavior, on dev it should be 0
+									CF_PAGES: '0',
+									// will be fetched from git dynamically once we support mocking of bindings
+									CF_PAGES_BRANCH: 'TBA',
+									// will be fetched from git dynamically once we support mocking of bindings
+									CF_PAGES_COMMIT_SHA: 'TBA',
+									CF_PAGES_URL: `http://${req.headers.host}`,
+									...vars,
+								},
+								cf: cf,
+								// will be supported once we support mocking of bindings
+								// waitUntil: (promise: Promise<any>) => {
+								//   context.waitUntil(promise);
+								// },
+								// will be supported once we support mocking of bindings
+								// caches: caches,
```

Comment:
I would think this can be mocked right here as well. Probably an in-memory cache, because caching across `astro dev` invocations is asking for trouble.

Reply:
Absolutely! The introduction of mocking is planned as a second phase, primarily because it necessitates more intricate architectural design. I aim to align this closely with the bindings feature.

As for caches, they will be excluded in the initial implementation. If you're dependent on cache functionality, the existing wrangler pages dev workaround remains a viable option for the time being.

Reply:
In that case, I think it makes sense for it be introduced as no-ops as well.